### PR TITLE
Add feature: premise retrieval

### PR DESCRIPTION
**Pull request by Peiyang-Song**
# Add feature: premise retrieval

This PR adds premise retrieval to Lean Copilot.

* [More info later]

### CODE REVIEW

File: cpp/ct2.cpp [hunk 1]
```diff
@@ -241,3 +247,129 @@ extern "C" lean_obj_res ct2_encode(b_lean_obj_arg _input_tokens) {
...
+  data_f.resize(data.size());
+  std::transform(data.begin(), data.end(), data_f.begin(),
+                 [](double d) { return static_cast<float>(d); });
+
+  std::vector<int64_t> shape_i64;
+  shape_i64.resize(shape.size());
+  std::transform(shape.begin(), shape.end(), shape_i64.begin(),
+                 [](unsigned long ul) { return static_cast<int64_t>(ul); });
+
+  premise_embeddings = new ctranslate2::StorageView(shape_i64, data_f, ctranslate2::Device::CPU);
```
Comment (yangky11) [inline cpp/ct2.cpp:276]: The device shouldn't be hardcoded.

Comment (Peiyang-Song) [reply cpp/ct2.cpp:287]: Just saw that we hardcoded device to be CPU when we were initializing encoder before. Would we like to change this as well?

Comment (yangky11) [reply cpp/ct2.cpp]: where?

Comment (Peiyang-Song) [reply cpp/ct2.cpp]: On line 218 (in the latest version) in `init_ct2_encoder`.

Comment (yangky11) [reply cpp/ct2.cpp]: I see. Can you change it to something similar to what we're doing for the generator?

Comment (Peiyang-Song) [reply cpp/ct2.cpp]: Sure!

Comment (Peiyang-Song) [reply cpp/ct2.cpp]: Done :)

**Pull request merged**